### ISSUE TITLE
Templating for global variables

### ISSUE THREAD

**Issue by phillipj** [tazjin/kontemplate#140]
Have you considered allowing global variables to support the same templating features as resource sets has?

My scenario would be something like this:

```yaml
---
context: k8s.cluster.com
global:
  version: {{ gitHEAD }}
include:
  - name: my-awesome-app

  ...
```

Mainly for convenience sake; making the default behaviour to use whatever is currently at `HEAD`, but still allow `--var version=an-explicit-sha` to be provided if needed.

**Comment by phillipj** [comment]
FYI this is what I'm getting when trying this in v1.6.0:

```
$ kontemplate template my-cluster.yaml
kontemplate: error: Error loading context: Context loading failed on file my-cluster.yaml due to:
error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{"gitHEAD":interface {}(nil)}
```

Another approach could to be to inline this in the resource sets, example psudeoish config;

```yaml
spec:
  containers:
    - image: xyz.ecr.eu-central-1.amazonaws.com/my-awesome-app:{{ getOrElse .version gitHEAD }}
      name: my-awesome-app
```

**Comment by tazjin** [comment]
Hey!

Due to how the context is deserialised it's not that easy to provide those at the top-level right now (Go was of course the wrong choice for this project as it doesn't allow useful generic traversal of data structures).

It could be done with some kind of extra key that is looked for specifically, but there are other concerns like evaluation order and whether a context interpolation even terminates etc.

TL;DR: It's possible, but I don't think I have capacity for it at the moment. If someone else wants to implement it, it'd be good to sanity-check the implementation with me first :)

In general I think it's easier to use things like the `default` function ([see here](http://masterminds.github.io/sprig/defaults.html)) in templates which basically does what your `getOrElse` mockup-call does.

**Comment by phillipj** [comment]
Oooh nice, thanks a bunch for the insight and protip about `default` 👍 If the latter does the trick, I'm more than happy to throw away the thought that demands a more complex solution of templating top-level vars.

**Comment by phillipj** [comment]
Getting an error when trying to use `default` since the `version` variable is not defined (when not provided as CLI arg):

```
Loading resources for my-awesome-app
kontemplate: error: Error templating resource sets: Error while templating my-awesome-app/config.yaml: template: config.yaml:14:96: executing "config.yaml" at <.version>: map has no entry for key "version"
```

I can't seem to find any go template or sprig functions that can help me check if the `version` is actually defined at all -- not just a empty or `nil` check.

Could it be reasonable to create a custom templater function in kontemplate that enables us to check for defined variables?

**Comment by tazjin** [comment]
> Getting an error when trying to use default since the version variable is not defined

Hmm, that's probably a conflict between [this setting](https://github.com/tazjin/kontemplate/blob/01f771fa2fb19ae6ca365c9d08000a81a266d609/templater/templater.go#L28) and the way the `default` function is implemented.

> Could it be reasonable to create a custom templater function in kontemplate that enables us to check for defined variables?

It may be possible to write an alternative default function that inspects the whole context and takes a key out of that (or alternatively a default value), without ever directly accessing the key via the template engine's standard mechanism.

I'd probably accept a PR for that, but don't currently have time to look at it myself.

**Comment by tazjin** [comment]
@phillipj Hey, is this issue still relevant for you or did the `default` function fix it?

**Comment by phillipj** [comment]
Yo! I think default covers my needs, thanks for much needed rubber ducking.
